### ISSUE TITLE
BERTopic fit_transform producing a different clustering than when invoking HDBSCAN.fit_predict directly on UMAP reductions

### ISSUE THREAD

**Issue by drob-xx** [MaartenGr/BERTopic#969]
Hi Maarten,

I'm having trouble understanding what is going on with the clustering of a model I'm working with. I've created a BERTopic instance with a UMAP and HDBSCAN (both cuML in this case) passed in when the instance is created. I then run BERTopic.fit_transform() and get a 14 topic clustering. However, the parameters I'm using for HDBSCAN in this case were meant to (and do) produce 11 clusters. While debugging I came across what seems like an inconsistency to me, but I'm wondering if I'm just missing something simple. This should illustrate the problem:

```
print(pd.Series(bt23543.hdbscan_model.labels_).value_counts())
hdbscan_model = deepcopy(bt23543.hdbscan_model)
hdbscan_model.fit_predict(bt23543.umap_model.embedding_)
print('')
print(pd.Series(hdbscan_model.labels_).value_counts()) 

-1     11254
 7      4865
 1      2012
 9      1498
 5      1043
 8       835
 6       667
 4       549
 11      543
 10      424
 3       404
 12      383
 2       255
 0       236
dtype: int64

-1     9368
 8     4926
 9     2371
 0     2226
 10    1902
 4     1103
 6      889
 7      706
 3      555
 2      419
 1      263
 5      240
dtype: int64
```
So I expected this output to be the same. Just to make sure that the original model actually used the same UMAP, HDBSCAN and setting I then re-ran and got this:

```
topics, probabilities = bt23543.fit_transform(ctmt.docs, embeddings=ctmt.embeddings)
pd.Series(bt23543.hdbscan_model.labels_).value_counts()

-1     11254
 7      4865
 1      2012
 9      1498
 5      1043
 8       835
 6       667
 4       549
 11      543
 10      424
 3       404
 12      383
 2       255
 0       236
dtype: int64
```

Why would BERTopic.fit_transform() wind up with a different clustering on the embeddings in this case than when I use hdbscan_model to cluster the UMAP embeddings? I'm stumped. Some more info:

- I'm running on colab+ with cuML set up. BERTopic is installed using pip install and is the latest version - 0.13.0
- Here is the BERTopic.get_params()

```
{'calculate_probabilities': True,
 'ctfidf_model': ClassTfidfTransformer(),
 'diversity': None,
 'embedding_model': <bertopic.backend._sentencetransformers.SentenceTransformerBackend at 0x7fc0506f4f40>,
 'hdbscan_model': HDBSCAN(),
 'language': None,
 'low_memory': False,
 'min_topic_size': 10,
 'n_gram_range': (1, 1),
 'nr_topics': None,
 'seed_topic_list': None,
 'top_n_words': 10,
 'umap_model': UMAP(),
 'vectorizer_model': CountVectorizer(ngram_range=(1, 3),
                 stop_words=frozenset({'a', 'about', 'above', 'across', 'after',
                                       'afterwards', 'again', 'against', 'all',
                                       'almost', 'alone', 'along', 'already',
                                       'also', 'although', 'always', 'am',
                                       'among', 'amongst', 'amoungst', 'amount',
                                       'an', 'and', 'another', 'any', 'anyhow',
                                       'anyone', 'anything', 'anyway',
                                       'anywhere', ...})),
 'verbose': False}
```
The HDBSCAN type - cuml.cluster.hdbscan.hdbscan.HDBSCAN
The HDBSCAN params:
```
{'handle': <pylibraft.common.handle.Handle at 0x7fc0506f4450>,
 'verbose': 4,
 'output_type': 'input',
 'metric': 'euclidean',
 'min_cluster_size': 235,
 'max_cluster_size': 0,
 'min_samples': 43,
 'cluster_selection_epsilon': 0.0,
 'cluster_selection_method': 'eom',
 'p': 2,
 'allow_single_cluster': False,
 'connectivity': 'knn',
 'alpha': 1.0,
 'gen_min_span_tree': None,
 'prediction_data': True}
```
The UMAP type - cuml.manifold.umap.UMAP
The UMAP params:
```
{'handle': <pylibraft.common.handle.Handle at 0x7fc0506f4390>,
 'verbose': 2,
 'output_type': 'input',
 'n_neighbors': 5,
 'n_components': 5,
 'n_epochs': 0,
 'learning_rate': 1.0,
 'min_dist': 0.0,
 'spread': 1.0,
 'set_op_mix_ratio': 1.0,
 'local_connectivity': 1.0,
 'repulsion_strength': 1.0,
 'negative_sample_rate': 5,
 'transform_queue_size': 4.0,
 'init': 'random',
 'a': 1.9328083978536088,
 'b': 0.7904949734413567,
 'target_n_neighbors': -1,
 'target_weight': 0.5,
 'target_metric': 'categorical',
 'hash_input': False,
 'random_state': 374756,
 'callback': None,
 'metric': 'cosine',
 'metric_kwds': None}
```

The difference could theoretically be caused by BERTopic not using hdbscan_model and using a non-cuML HDBSCAN instead. I poked through the code and I do see a couple of times where you test for hdbscan.HDBSCAN and or use a new HDBSCAN instance (as in `_auto_reduce_topics`) and wondered if this could possibly lead to the problem - using a different HDBSCAN implementation could/would certainly cause this behavior. However, I don't see clearly where this might be occuring. As always thanks in advance.

**Comment by drob-xx** [comment]
So it seems to be a problem when using cuML UMAP - as illustrated here:

```
from sklearn.datasets import fetch_20newsgroups
from cuml.manifold import UMAP
from hdbscan import HDBSCAN
from bertopic import BERTopic

docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data'][:200]
umap_model = UMAP(
                n_components=5,
                metric="cosine",
                n_neighbors=5,
                min_dist=0.0,
                verbose=2,
                random_state=np.uint64(42),
                init="random",
            )
bt23543 = BERTopic(umap_model=umap_model)
bt23543.fit_transform(docs)
print(pd.Series(bt23543.hdbscan_model.labels_).value_counts())
hdbscan_model = HDBSCAN(**bt23543.hdbscan_model.get_params())
hdbscan_model.fit(bt23543.umap_model.embedding_)
print('')
print(pd.Series(hdbscan_model.labels_).value_counts())
```

Which outputs:

```
 2    73
-1    53
 1    51
 0    23
dtype: int64

 0    86
-1    53
 3    24
 1    22
 2    15
dtype: int64
```

**Comment by drob-xx** [comment]
Hi @MaartenGr I know you are busy. In the interest of making this as straightforward as possible I have another, even easier and more direct test that shows something isn't right. By simply changing the UMAP implementation to cuML we get a behavior from BERTopic that I don't think is correct. I've looked pretty deeply into the BERTopic code and this has me completely stumped. If it is a cuML problem I would like to open an issue there - but right now I can't reproduce without BERTopic and I'm sure they will want to see the problem more clearly. Here's the test case:

```
from sklearn.datasets import fetch_20newsgroups
from hdbscan import HDBSCAN
from bertopic import BERTopic
import pandas as pd

docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data'][:200]

def runTest(docs):
  umap_model = UMAP(
                  n_components=5,
                  metric="cosine",
                  n_neighbors=5,
                  min_dist=0.0,
                  # verbose=2,
                  # random_state=np.uint64(42),
                  # init="random",
              )
  btModel = BERTopic(umap_model=umap_model, verbose=0)
  btModel.fit_transform(docs)
  print("Summarize BERTopic topics_")
  print(pd.Series(btModel.hdbscan_model.labels_).value_counts())
  print("Summarize HDBSCAN labels_")
  print(pd.Series(btModel.hdbscan_model.labels_).value_counts())
  print('Recluster manually, summarize HDBSCAN labels_')
  print(pd.Series(btModel.hdbscan_model.fit_predict(btModel.umap_model.embedding_)).value_counts())
  return btModel

from umap import UMAP
print('')
print('---------------------------------------')
print('Running with umap')
umap_bt = runTest(docs)

from cuml.manifold import UMAP
print('---------------------------------------')
print('Running with cuml')
cuml_bt = runTest(docs)

```
And my output:

```

---------------------------------------
Running with umap
Summarize BERTopic topics_
 2    87
 0    61
-1    30
 1    22
dtype: int64
Summarize HDBSCAN labels_
 2    87
 0    61
-1    30
 1    22
dtype: int64
Recluster manually, summarize HDBSCAN labels_
 2    87
 0    61
-1    30
 1    22
dtype: int64
---------------------------------------
Running with cuml
Summarize BERTopic topics_
 0    74
-1    61
 2    48
 1    17
dtype: int64
Summarize HDBSCAN labels_
 0    74
-1    61
 2    48
 1    17
dtype: int64
Recluster manually, summarize HDBSCAN labels_
 2    76
 0    49
-1    48
 1    27
dtype: int64

```

**Comment by MaartenGr** [comment]
When you are devising these kinds of test cases, it might be worthwhile to also check whether the inputs to the models are exactly the same. Here, and this seems to be what the issue is, you assume that the reduced embeddings of cuML's UMAP are exactly the same as `umap_model.embedding_` which is not the case:

```python
import numpy as np
from sklearn.datasets import fetch_20newsgroups
from sentence_transformers import SentenceTransformer

from umap import UMAP
from cuml.manifold import UMAP as cuml_UMAP

# Prepare embeddings
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data'][:200]
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = sentence_model.encode(docs, show_progress_bar=False)

# Prepare models
cuml_umap_model = cuml_UMAP(n_components=5, metric="cosine", n_neighbors=5, min_dist=0.0, random_state=np.uint64(42), init="random")
umap_model = UMAP(n_components=5, metric="cosine", n_neighbors=5, min_dist=0.0, random_state=np.uint64(42), init="random")

# UMAP
umap_model.fit(embeddings)
reduced_embeddings = umap_model.transform(embeddings)
assert np.array_equal(umap_model.embedding_, reduced_embeddings)  # Holds True

# cuML UMAP .fit_transform
reduced_embeddings = cuml_umap_model.fit_transform(embeddings)
assert np.array_equal(cuml_umap_model.embedding_, reduced_embeddings)  # Holds True

# cuML UMAP .fit then .transform
cuml_umap_model.fit(embeddings)
reduced_embeddings = cuml_umap_model.transform(embeddings)
assert np.array_equal(cuml_umap_model.embedding_, reduced_embeddings)  # Raises AssertionError
```

This is not actually a bug as it has to do with the `hash_input` [parameter](https://github.com/rapidsai/cuml/blob/f6abf81ac64b805498ee5ebbaeedd8fb4d5b16cf/python/cuml/manifold/umap.pyx#L231-L240) of cuML's UMAP. Taking a snippet of that docstring:

> This enables consistent
        behavior between calling ``model.fit_transform(X)`` and
        calling ``model.fit(X).transform(X)``. Not that the CPU-based
        UMAP reference implementation does this by default. This
        feature is made optional in the GPU version due to the
        significant overhead in copying memory to the host for
        computing the hash.

In other words, when you set `hash_input=True` when instantiating cuML's UMAP, you will not get an assertion error:

```python

# Prepare models
cuml_umap_model = cuml_UMAP(n_components=5, metric="cosine", n_neighbors=5, min_dist=0.0, random_state=np.uint64(42), init="random",
                            hash_input=True)
cuml_umap_model.fit(embeddings)
reduced_embeddings = cuml_umap_model.transform(embeddings)
assert np.array_equal(cuml_umap_model.embedding_, reduced_embeddings)  # Holds True
```

**Comment by drob-xx** [comment]
Thanks once again so much Maarten. After reading through the code I was pretty sure that it wasn't a BERTopic issue, but I just didn't know where to look. The `hash_input=True` worked of course. I have opened an [issue](https://github.com/rapidsai/cuml/issues/5188) with rapidsai and I'm curious to see their response. Maybe I'm missing something but it seems to me that it would have been better to default this value to `True` and allow people to optionally set it to `False`. I'm not sure in what world one would not expect `embedding_` to actually return the embedding :), especially since it does in the umap-learn version. But we'll see what they say and it wouldn't be the first time I didn't understand something. Thanks!
